proposals/accepted/051-ghc-base-libraries.rst

```diff
@@ -0,0 +1,331 @@
+   (Incidentally, ``base`` and ``ghc-prim`` are both part of the same GitHub repository, which also includes GHC's source code.)
+
+Proposal
+========
+
+We propose to divide ``base`` into three packages:
+
+- ``ghc-internals``: exposes aspects of GHC's internals that may be of interest to "hard-core" developers interested in maximum performance (see `Nikita's blog post <https://nikita-volkov.github.io/internal-convention-is-a-mistake/>`__).
+  The API of ``ghc-internals`` is fully under the control of the GHC team, and of no direct interest to the CLC — only its effects on the API of base.
+
+- ``base``: as now, whose API is curated by CLC.
+  Depends on ``ghc-internals``, and hence on ``ghc-bignum`` and ``ghc-prim``.
+
+- ``ghc-experimental``, initially empty, depends on ``base`` and on ``ghc-internals``.
+  Functions and data types here are intended to have their ultimate home in base, but while they are settling down they are subject to much weaker stability guarantees.
+  Generally, new functions and types introduced in GHC Proposals would start their life here.
+  Example: new type families and type constructors for tuples, `GHC Proposal #475 <https://github.com/ghc-proposals/ghc-proposals/pull/475>`__.
+
+  Another example: future APIs to access RTS statistics, which are fairly stable and user-exposed, but which are (by design) coupled closely to GHC's runtime and hence may change.
+
+  As its name suggests, the API of ``ghc-experimental`` is curated by the GHC team, although the CLC is willing to offer (non-binding) opinions, if consulted.
+
+All three packages conform rigorously to the PVP.
+(But see Section 5.3)
+
+Some observations about this structure:
+
+- We should use all possible social and technical means to discourage people from depending directly on ``ghc-internals``, because if such dependencies become frequent and ossified, it will lead to future pain when the API changes.
+  Saying "we told you not to rely on it" may be true, but won't lessen that pain.
+  What mechanisms could we use?
+
+  - The name ``ghc-internals`` is a pretty strong signal all by itself.
+
+  - Cabal description and README explains how it is intended used (and not used).
+
+  - Hoogle could (by default anyway) never show stuff from ``ghc-internals``.
+
+  - Do not upload Haddocks for ``ghc-internals`` to Hackage.
+    (Ditto ``ghc-prim``.) Need to make sure that if someone wants to follow the Haddock source-code link to (say) Functor, they should still find it regardless of where it is actually defined.
+
+  - We could consider issuing a warning if you say ``-package ghc-internals`` (or ``ghc-bignum`` or ``ghc-prim``), one that was hard to silence.
+    Since we can have module-level ``WARNING`` pragmas with custom categories, one way to realise this would be to pick a category and add such pragmas to every module in the relevant packages, though we might want to do something more systematic.
```

Comment:
This sounds excellent. Much like "partial", "internal" seems like a fairly natural warning category that other people in the ecosystem might want to use. And it's easy for a user to say "I know what I'm doing" by just turning off that warning category in the module where they use internals.

Reply:
There's the `category` field in cabal files which is pretty unstructured an all over the map. We could easily initiate a convention of the `internal` category actually meaning something -- e.g. triggering an extra textual notice on hackage pages listing the package, and extending cabal so it could maybe optionally filter or warn on such stuff.

Reply:
That seems like more work for less fine control? If we already have the mechanism in GHC at a module level, that seems sufficient to me.

Reply:
Hackage itself cannot and should not read module-level pragmas, nor can cabal. This proposal is about having some _packages_ be internal and some not. So we need package-level mechanisms for marking things internal, such that both hackage and cabal can take appropriate action (perhaps additional html, like a big red warning box on hackage, and for cabal the ability to warn on using internals as a _direct_ rather than indirect dependency, etc)

Reply:
> and for cabal the ability to warn on using internals as a direct rather than indirect dependency, etc

From what I understood people are also worried about indirect dependencies (coming from things other than the original package), because they may easily lack behind upgrades and bubble up churn.

Reply:
Yes, that would be another thing that would be nice to address. the `category` field does not address that, but seems useful nonetheless, and I would encourage its consideration regardless.

Reply:
(Sorry, I realised that what I was saying was ambiguous based on the line I was commenting on: I'm in favour of a module-level INTERNAL warning, _added manually to source files_, not magically added to packages by cabal or Hackage)